(experimental) Add Addressee class.

showell opened the pull request.

smarx commented:
*Automated message from Dropbox CLA bot*

@showell, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -99,6 +101,24 @@ def test_creating_streams(self):
+            welcome_message,
+            u'Welcome to this stream! Please say hello.\n\nThis is a stream'
+        )
+
```
We should have a test for the no-description case too.

timabbott commented on zerver/views/streams.py:
```diff
@@ -283,22 +290,41 @@ def add_subscriptions_backend(request, user_profile,
+                        content=msg))
+
+    for stream in created_streams:
+        notifications.append(prep_stream_welcome_message(stream))
```
We may want an `if not realm.is_zephyr_mirror_realm` check here; .

timabbott replied:
(Basically because in the Zephyr or more generally mirror world, creating a stream in Zulip may be actually just creating a reflection of an already-existing thing outside Zulip).

showell commented:
I pushed out new changes to address the comments made on this PR.

There was some discussion on chat about how to format the welcome message, and I addressed some of those as well, but I think some of the ideas can be incorporated in future iterations.

I'm happy to abandon the last commit in this series, if it's not to your taste.  Adding the Addressee class does simplify a few things, but it's not as big of an overall win as I had hoped for, so it might not be worth the overhead of the new class, plus we're getting close to doing a release.

timabbott commented:
Merged all but the last commit.  

For the last commit, I think it's a nice instantiation of the idea, but I worry that may add more complexity than it's worth; especially since we do have mypy providing a bit of type safety in that code path already.  But I'm not sure.

showell commented:
Ok, for now I rebased the PR so it's just the unmerged commit, and I changed the title of the PR.

If somebody stumbles on this after Spring 2017, they can consider this a failed experiment and close the PR.
